functions/src/index.ts

```diff
@@ -0,0 +1,231 @@
+            name: gameName,
+            password: password,
+            genre: gameGenre,
+            rounds: roundDetails,
+            intermissionDuration: NaN,
+            roundDuration: NaN,
+            startTime: null,
+            leaderBoard: {
+                [hostId]: {
+                    name: hostName,
+                    score: 0
+                }
+            }
+        }
+        admin.firestore().collection("games").add(gameDetails).then(entry => {
+            response.send(entry.id)
+            return entry
+        }).catch(err => {
+            console.log(err)
+            throw err
+        })
+        return gameDetails
+    }).catch(error => {
+        console.log(error)
+        response.send(error)
+        return error
+    })
+})
+
+function getRoundDetails(gameGenre: string, gameRounds: number, optionsPerRound: number) {
+    return getSpotifyToken().then(res => {
+        const access_token = res
+        return getSpotifyTracks(gameGenre, gameRounds, access_token, optionsPerRound).then((tracks: SpotifyTrack[]) => {
+            const answerTracks: GameTrack[] = []
+            const answerUrls: string[] = []
+
+            // there is a chance that there will not be enough tracks with previews
+            // TODO: add fallback logic to get more tracks with previews
+            for (const track of tracks) {
+                console.log(track)
+                if (track.preview_url !== null) {
+                    answerUrls.push(track.preview_url)
+                    answerTracks.push({
+                        id: track.id,
+                        title: String(track.name),
+                        artists: track.artists.map(artist => artist.name),
+                        isAnswer: true,
+                    })
+                }
+                if (answerUrls.length === gameRounds) {
+                    break
+                }
+            }
+
+            //making round objects
+            const gameRoundDetails: RoundDetails[] = []
+            for (let i = 0; i < gameRounds; i++) {
+                const gameTracks: GameTrack[] = []
+                gameTracks.push(answerTracks[i])
+                for (let j = 0; j < tracks.length; j++) {
+                    // Make sure we don't add the answer track
+                    if (tracks[j].id === answerTracks[i].id) {
+                        continue
+                    }
+
+                    gameTracks.push({
+                        id: tracks[j].id,
+                        title: tracks[j].name,
+                        artists: tracks[j].artists.map(artist => artist.name),
+                        isAnswer: false
+                    })
+
+                    if (gameTracks.length === optionsPerRound) {
+                        break
+                    }
+                }
+
+                gameRoundDetails.push({
+                    url: answerUrls[i],
+                    tracks: gameTracks
+                })
+            }
```

Comment:
Having a hard time following those 2 loops. I don't think we should iterate over gameRounds (`for (let i = 0; i < gameRounds; i++)`) and using the gameRound index to access answerTracks (`answerTracks[i]`), it seems very error prone to me. However if this works, fuck it. Maybe we can think about this later if needed.

Reply:
yeah I agree, kinda sketchy, but my comment above explains why i need to first find the answer tracks